doc/GUIDE.md

```diff
@@ -0,0 +1,822 @@
+```
+
+The `MyPoint` `typedef` on the other hand is a typedef for a struct, so it is translated
+as a separate class which extends the class that is generated for the `Point` struct:
+
+```java
+// MyPoint.java
+
+public class MyPoint extends Point { }
+```
+
+Through static inheritance, all the methods in the `Point` class are available through the
+`MyPoint` class as well.
+
+### Nested Types
+
+C allows variable declarations to have an inline anonymous type. Jextract handles in
+particular cases where a struct's field has an inline type specially. For instance, if we
+have a struct such as this:
+
+```c
+// mylib.h
+
+struct Foo {
+    struct {
+        int baz;
+    } bar; // field of Foo
+
+    void (*cb)(void);
+};
+```
+
+Jextract generates a _nested_ struct and function pointer class for the `bar` and `cb`
+fields, _inside of_ the class it generates for the `Foo` struct itself:
+
+```java
+// mylib_h.java
+
+public class Foo {
+    ...
+    public static class bar { ... }
+    public static final GroupLayout bar$layout() { ... }
+    public static final long bar$offset() { ... }
+
+    public static MemorySegment bar(MemorySegment struct) { ... }
+    public static void bar(MemorySegment struct, MemorySegment fieldValue) { ... }
+    ...
+    public static class cb { ... }
+    public static final AddressLayout cb$layout() { ... }
+    public static final long cb$offset() { ... }
+
+    public static MemorySegment cb(MemorySegment struct) { ... }
+    public static void cb(MemorySegment struct, MemorySegment fieldValue) { ... }
+    ...
+}
+```
+
+In both cases, the name of the nested class is the name of the field of the struct.
+
+Both fields also have the usual getter and setter, but note that the getter for the struct
+field returns a _reference_ to the memory inside the `Foo` struct that corresponds to the
+`bar` field. This means that writes to the returned memory segment will be visible in the
+enclosing struct instance as well:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    MemorySegment foo = Foo.allocate(arena);
+    MemorySegment bar = Foo.bar(foo); // reference to bar inside foo
+    System.out.println(Foo.bar.baz(bar));  // prints: 0
+
+    MemorySegment bar2 = Foo.bar.allocate(arena);
+    Foo.bar.baz(bar2, 42);
+    Foo.bar(foo, bar2);  // copies bar2 into foo
+    System.out.println(Foo.bar.baz(bar));  // prints: 42
+}
+```
+
+In the above snippet, note that the load of the `baz` field value on the last line will
+'see' the update to the `bar` field of the `foo` instance on the line before.
+
+## Filtering
+
+Some libraries are incredibly large (such as a platform SDK), and we might not be
+interested in letting jextract generate code for the entire library. In cases like that,
+we can use jextract's `--include-XXX` command line options to only generate classes for
+the elements we specify.
+
+To allow for symbol filtering, `jextract` can generate a _dump_ of all the symbols
```

Comment:
I noticed that in these sections, refrerences to `jextract` use a code block, whereas in the rest of the document they don't. I don't have a strong preference for one or the other, but we should make that consistent.

Reply:
I'll just switch to the plain jextract, as that's a little easier on the eyes. (This section was copied from the readme)